tests/chainer_tests/links_tests/normalization_tests/test_batch_normalization.py

```diff
@@ -36,118 +39,175 @@ def _batch_normalization(expander, gamma, beta, x, mean, var, eps, test):
-            x = chainer.Variable(x_data)
-            y = self.link(x)
-            self.assertEqual(y.data.dtype, self.dtype)
-
-        y_expect = _batch_normalization(
-            self.expander, self.gamma, self.beta, self.x, self.mean,
-            self.var, self.link.eps, self.test)
+            y = link(x, finetune=self.finetune)
+        return y,
 
-        testing.assert_allclose(
-            y_expect, y.data, **self.check_forward_optionss)
-
-    @condition.retry(3)
-    def test_forward_cpu(self):
-        self.check_forward(self.x)
+    def forward_expected(self, inputs, params):
+        x, = inputs
+        gamma, beta = params
 
-    @attr.gpu
-    @condition.retry(3)
-    def test_forward_gpu(self):
-        self.link.to_gpu()
-        self.check_forward(cuda.to_gpu(self.x))
-
-    @attr.cudnn
-    def test_forward_gpu_without_cudnn(self):
-        with chainer.using_config('use_cudnn', 'never'):
-            self.test_forward_gpu()
-
-    @attr.multi_gpu(2)
-    @condition.retry(3)
-    def test_forward_multi_gpu(self):
-        with cuda.get_device_from_id(1):
-            self.link.to_gpu()
-            x = cuda.to_gpu(self.x)
-        with cuda.get_device_from_id(0):
-            self.check_forward(x)
-
-    def check_backward(self, x_data, y_grad):
-        gradient_check.check_backward(
-            self.link, x_data, y_grad, (self.link.gamma, self.link.beta),
-            eps=1e-2, **self.check_backward_optionss)
-
-    @condition.retry(3)
-    def test_backward_cpu(self):
-        self.check_backward(self.x, self.gy)
-
-    @attr.gpu
-    @condition.retry(3)
-    def test_backward_gpu(self):
-        self.link.to_gpu()
-        self.check_backward(cuda.to_gpu(self.x), cuda.to_gpu(self.gy))
-
-    @attr.cudnn
-    def test_backward_gpu_without_cudnn(self):
-        with chainer.using_config('use_cudnn', 'never'):
-            self.test_backward_gpu()
+        if self.test:
+            mean = self.mean[self.expander]
+            var = self.var[self.expander]
+            std = numpy.sqrt(var)
+        else:
+            mean = x.mean(axis=self.aggr_axes, keepdims=True)
+            var = x.var(axis=self.aggr_axes, keepdims=True)
+            std = numpy.sqrt(var + self.eps)
+        y = gamma[self.expander] * (x - mean) / std + beta[self.expander]
+        return y,
+
+
+@testing.inject_backend_tests(
+    None,
+    # CPU tests
+    [
+        {},
+        {'use_ideep': 'always'},
+    ]
+    # GPU tests
+    + testing.product({
+        'use_cuda': [True],
+        'use_cudnn': ['never', 'always'],
+        'cuda_device': [0, 1],
+    })
+    # ChainerX tests
+    + [
+        {'use_chainerx': True, 'chainerx_device': 'native:0'},
+        {'use_chainerx': True, 'chainerx_device': 'cuda:1'},
+        {'use_chainerx': True, 'chainerx_device': 'cuda:1'},
+    ])
+@_parameterize
+class BatchNormalizationTest(BatchNormalizationTestBase, testing.LinkTestCase):
+
+    contiguous = 'C'
```

Comment:
Is this needed?

Reply:
Thanks, it was there during debugging. Removed.